### ISSUE TITLE
Revisit the timeslicing algorithm

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#855]
**_Reported by Benjamin Walsh:_**

The timeslicing algorithm currently does not reset the remaining timeslice when swapping to a new thread of higher priority. Revisit this.

(Imported from Jira ZEP-948)

**Comment by nashif** [comment]
**_by Andy Ross:_**

Will investigate.  I assume the intended behavior is that we should get a new full timeslice on priority elevation?

FWIW: does that really rise to a high priority bug?  Seems like the impact is a pretty weird situation: you need a situation where you have two "high" priority tasks that are CPU-bound and competing with each other for time slices, *and* where only one of the two is routinely blocked on a lower priority thread that will switch to it, and that lower priority thread needs to be running and taking up significant parts of the timeslice (i.e. the other high priority thread needs to be blocked too!) so that it's "stealing" some from the high priority thread.

Basically to exercise this, everyone needs to be rapidly blocking/unblocking *and* still depend on fair timeslicing of CPU resources.  I'm not even sure how I'd write a test case to exercise this if I tried.

**Comment by nashif** [comment]
**_by Ramesh Thomas:_**

This now gets done in the Tickless Kernel implementation.

**Comment by nashif** [comment]
**_by Sharron LIU:_**

https://github.com/zephyrproject-rtos/zephyr/pull/80
_Ramesh Thomas_ , could you review the test case?

The test run with 4 preemptive thread in same priority, with timeslice enabled.

a ztest native thread (taken a half timeslice, then relinquish CPU)
thread_1 (expect taking a full timeslice, however the test result shows it took a half timeslice)
thread_2 (expect taking a full timeslice)
thread_3

The test result I observe is thread_1 actually took a "half" timeslice. Could you review if I understand the expected behavior properly?

NOTE: the test case did not enable tickless idle. Reason is, from the story description, timeslice behavior should keep consistent regardless whether or not tickless idle is enabled. What's more, the case failed even with tickless idle enable.

Issue tracked as GH-1948

**Comment by nashif** [comment]
**_by Ramesh Thomas:_**

The time slicing algorithm is meant to only ensure that the thread does not execute more than the time slice duration if there are multiple threads with same priority and they meet all conditions necessary for time slicing.

Did you check if there were any events that caused the swap of threads? It is expected to swap out before the time slice if there is an interruption.

**Comment by nashif** [comment]
**_by Ramesh Thomas:_**

_Antonino Scarpaci_   why have you changed the status to "code review"?  "merged" is the correct status for this story. Please open a new Jira bug if you see any issue.

**Comment by nashif** [comment]
**_by Antonino Scarpaci:_**

Sorry, I apologize for my error while I read the timeslice feature

**Comment by nashif** [comment]
**_by Sharron LIU:_**

I think we should move it back to Merged?

**Comment by nashif** [comment]
**_by Sharron LIU:_**

Issue tracked as GH-1948, closing this.
